src/BinOp.chpl

```diff
@@ -0,0 +1,255 @@
+  :type rname: string
+
+  :arg pn: routine name of callsite function
+  :type pn: string
+
+  :arg st: SymTab to act on
+  :type st: borrowed SymTab 
+
+  :returns: (MsgTuple) 
+  :throws: `UndefinedSymbolError(name)`
+  */
+  proc doBinOpvv(l, r, e, op: string, rname, pn, st) throws {
+    // Since we know that the result type is a boolean, we know
+    // that it either (1) is an operation between bools or (2) uses
+    // one of the boolean operators from the `boolOps` set above
+    if e.etype == bool {
+      if l.etype == bool && r.etype == bool {
+        select op {
+          when "|" {
+            e.a = l.a | r.a;
+          }
+          when "&" {
+            e.a = l.a & r.a;
+          }
+          when "^" {
+            e.a = l.a ^ r.a;
+          }
+          when "==" {
+            e.a = l.a == r.a;
+          }
+          when "!=" {
+            e.a = l.a != r.a;
+          }
+          otherwise {
+            var errorMsg = notImplementedError(pn,l.dtype,op,r.dtype);
+            omLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
+            return new MsgTuple(errorMsg, MsgType.ERROR);
+          }
+        }
+      }
+      // All types support the same binary operations when the resultant
+      // type is bool and `l` and `r` are not both boolean, so this does
+      // not need to be specialized for each case.
+      else {
+        select op {
+            when "<" {
+              e.a = l.a < r.a;
+            }
+            when ">" {
+              e.a = l.a > r.a;
+            }
+            when "<=" {
+              e.a = l.a <= r.a;
+            }
+            when ">=" {
+              e.a = l.a >= r.a;
+            }
+            when "==" {
+              e.a = l.a == r.a;
+            }
+            when "!=" {
+              e.a = l.a != r.a;
+            }
+            otherwise {
+              var errorMsg = notImplementedError(pn,l.dtype,op,r.dtype);
+              omLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);                              
+              return new MsgTuple(errorMsg, MsgType.ERROR); 
+            }
+          }
+      }
+      var repMsg = "created %s".format(st.attrib(rname));
+      return new MsgTuple(repMsg, MsgType.NORMAL);
+    }
+    // Since we know that both `l` and `r` are of type `int` and that
+    // the resultant type is not bool (checked in first `if`), we know
+    // what operations are supported based on the resultant type
+    else if l.etype == int && r.etype == int {
+      if e.etype == int {
+        select op {
+          when "+" {
+            e.a = l.a + r.a;
+          }
+          when "-" {
+            e.a = l.a - r.a;
+          }
+          when "*" {
+            e.a = l.a * r.a;
+          }
+          when "//" { // floordiv
+            ref ea = e.a;
+            ref la = l.a;
+            ref ra = r.a;
+            [(ei,li,ri) in zip(ea,la,ra)] ei = if ri != 0 then li/ri else 0;
+          }
```

Comment:
I accidentally didn't save one of my comments, but it's about existing code and not new changes so kinda outside the scope of the PR anyway

No change requested:
I was thinking it might be nice to add a comment about the `ref`s and `zip`s for the first of these

Also do we need `ref`s of `l.a` and `r.a`? My (limited) understanding is `e.a` is the only one that needs a `ref` since it's the only one being updated. Would something like below work just as well or am I missing something?
```C
when "//" { // floordiv
  ref ea = e.a;
  [(ei,li,ri) in zip(ea,l.a,r.a)] ei = if ri != 0 then li/ri else 0;
}
```

Reply:
@bmcdonald3 @pierce314159 @bradcray the ref is required here because the array is wrapped in an object and the object is stored on locale-0. the refs are not required when you do a direct whole array operations like `e.a = l.a + r.a`, i think this is because the compiler can infer that the object references are constant through the whole operation.

Reply:
@pierce314159 : For reference, this is something that's required for performance, not correctness, and which we hope to have the Chapel compiler automatically optimize in the future—so view it as an unfortunate workaround rather than anything we're particularly proud of or that should obviously make sense.  The following issue in our repo tracks it:  https://github.com/chapel-lang/chapel/issues/10160